### PR TITLE
Homework

### PR DESCRIPTION
**Pull request by Kuba-Wi**
Fixed acceleration for values less than zero
Deleting engines in cars' destructors
Can't change engine when speed != 0
And some other small stuff

### CODE REVIEW

File: ElectricCar.cpp [hunk 1]
```diff
@@ -7,10 +7,21 @@ ElectricCar::ElectricCar(ElectricEngine* engine)
...
+    std::cout << __FUNCTION__ << std::endl;
+    delete engine_;
+}
 void ElectricCar::charge()        { std::cout << __FUNCTION__ << std::endl; }
 
+void ElectricCar::restore() {
+    charge();
+}
+
+void ElectricCar::changeEngine(int power, int batteryCapacity) {
```
Comment (WBigaj) [inline ElectricCar.cpp:20]: Isn't better to just take engine as parameter?

**Review by WBigaj — COMMENTED** [review]
I have checked your code and have one question?

File: Car.cpp [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+void Car::turnRight()     { std::cout << __FUNCTION__ << std::endl; }
+void Car::brake() {
+    std::cout << __FUNCTION__ << std::endl;
+    speed_ = 0;
+}
+void Car::accelerate(int speed) {
+    if(speed > 0) {
+        std::cout << __FUNCTION__ << std::endl;
+        speed_ = speed;
+    }
```
Comment (MaciejSalamonski) [inline Car.cpp:15]: What do you think about making an exception here? Ofc when speed <=0

File: ElectricCar.cpp [hunk 3]
```diff
@@ -7,10 +7,21 @@ ElectricCar::ElectricCar(ElectricEngine* engine)
...
+    if(speed_ != 0) {
+        return;
+    }
```
Comment (MaciejSalamonski) [inline ElectricCar.cpp:23]: Maybe lets throw sth here.

File: PetrolEngine.cpp [hunk 4]
```diff
@@ -5,16 +5,16 @@ PetrolEngine::PetrolEngine(int power, float capacity, int gears)
...
 {
     std::cout << __FUNCTION__ << std::endl;
 }
 
 void PetrolEngine::changeGear(int gear)
 {
-    // TODO: Add checking if gear is between -1 and gears_
-    // -1 is for REAR
-    // 0 is for NEUTRAL
+    if(gear < -1 || gear > gears_ || (currentGear_ != 0 && gear == -1)) {
```
Comment (MaciejSalamonski) [inline PetrolEngine.cpp:15]: I suggest you to avoid magic numbers and define variables like -1 as constexpr or const variables

**Review by MaciejSalamonski — APPROVED** [review]
Hi! good job ;). Have a good day!

File: ElectricCar.cpp (hunk 1 above)
Comment (Kuba-Wi) [reply ElectricCar.cpp:20]: Don't know if better but in different way.

File: PetrolEngine.cpp (hunk 4 above)
Comment (Kuba-Wi) [reply PetrolEngine.cpp:15]: Agree

File: Car.cpp (hunk 2 above)
Comment (Kuba-Wi) [reply Car.cpp:15]: Why not

File: ElectricCar.cpp (hunk 3 above)
Comment (Kuba-Wi) [reply ElectricCar.cpp:23]: Maybe

File: ElectricCar.cpp (hunk 1 above)
Comment (ziobron) [reply ElectricCar.cpp:20]: Passing engine would be a desired way. Then we can have something like EngineFactory that can produce engines and we can pass them to this function. However, for current needs it is ok.

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
4 XP za wyjątek InvalidGear w Cars (3521)
2 * 2 XP za poprawkę interfejsu w Cars (3522)
2 * 2 XP za punktualność (1119)
{ "requirement_ids": [3521, 3522, 3522, 1119, 1119] }

**Pull request closed without merging**